core/src/banking_stage/transaction_scheduler/multi_iterator_consume_scheduler.rs

```diff
@@ -0,0 +1,617 @@
+        if !self
+            .batch_account_locks
+            .check_sanitized_message_account_locks(message)
+        {
+            return ConsumeSchedulingDecision::Later;
+        }
+
+        // Check if this transaction takes locks that conflict with any transactions
+        // which were passed over due to above check.
+        if !self
+            .batch_priorty_order_guard
+            .check_sanitized_message_account_locks(message)
+        {
+            return ConsumeSchedulingDecision::Later;
+        }
+
+        // Check if this transaction takes locks that conflict with any transactions
+        // that were unschedulable due to below check.
+        if !self
+            .pass_priority_order_guard
+            .check_sanitized_message_account_locks(message)
+        {
+            return ConsumeSchedulingDecision::NextPass;
+        }
+
+        let transaction_locks = transaction.get_account_locks_unchecked();
+        if let Some(thread_id) = self.account_locks.try_lock_accounts(
+            transaction_locks.writable.into_iter(),
+            transaction_locks.readonly.into_iter(),
+            self.schedulable_threads,
+            |thread_set| {
+                Self::select_thread(
+                    &self.batches.transactions,
+                    self.in_flight_tracker.num_in_flight_per_thread(),
+                    thread_set,
+                )
+            },
+        ) {
+            ConsumeSchedulingDecision::Add(thread_id)
+        } else {
+            ConsumeSchedulingDecision::NextPass
+        }
+    }
+
+    fn add_transaction_to_batch(
+        &mut self,
+        id: &TransactionPriorityId,
+        SanitizedTransactionTTL {
+            transaction,
+            max_age_slot,
+        }: SanitizedTransactionTTL,
+        cu_limit: u64,
+        thread_id: ThreadId,
+    ) {
+        self.batch_account_locks
+            .add_sanitized_message_account_locks(transaction.message());
+        self.batches.transactions[thread_id].push(transaction);
+        self.batches.ids[thread_id].push(id.id);
+        self.batches.max_age_slots[thread_id].push(max_age_slot);
+        self.batches.total_cus[thread_id] += cu_limit;
+
+        if self.batches.ids[thread_id].len()
+            + self.in_flight_tracker.num_in_flight_per_thread()[thread_id]
+            >= QUEUED_TRANSACTION_LIMIT
+        {
+            self.schedulable_threads.remove(thread_id);
+        }
+    }
+
+    fn select_thread(
+        batches_per_thread: &[Vec<SanitizedTransaction>],
+        in_flight_per_thread: &[usize],
+        thread_set: ThreadSet,
+    ) -> ThreadId {
+        thread_set
+            .contained_threads_iter()
+            .map(|thread_id| {
+                (
+                    thread_id,
+                    batches_per_thread[thread_id].len() + in_flight_per_thread[thread_id],
+                )
+            })
+            .min_by(|a, b| a.1.cmp(&b.1))
```

Comment:
So rn it is load balancing by number of transactions per thread? Not super confident if doing so is necessary better for performance, i mean for what we have now, 1 batch of 64 TXs should be better than 2 batches of 32 txs, is it?

Reply:
If we only have 64 transactions, 2 batches of 32 txs would almost certainly be better.
Then we are executing them in parallel.
It's less efficient for PoH, since we'll lock and add 2 entries, but the bottleneck is mainly execution

Reply:
yea, I was thinking about poh locking - poh is *the* critical resource, should try the best _not_ to lock it. Agree in this example, two threads should out perform one additional locking. 

I was wondering if necessary to invest more in finding alternatives other than round robin. But It is a good start, should let perf data speak :)

Reply:
If we have enough txs to fill N batches (N being number of threads), we will end up recording `N` hashes into PoH.
I think the only case where we'd end up recording more hashes is when we don't have enough transactions, in which case I'd argue we don't need to care about the locking as much?